### PR TITLE
add retinanet with resnet backbone

### PR DESCRIPTION
**Pull request by sohaiberrabii**
- adds resnext to support loading retinanet pretrained on openimages
- object detection post processing with numpy. Some parts could easily be written with tinygrad ops but im not sure how useful that would be. imo it should stay decoupled from forward until the whole postprocessing is part of the graph.
- Just like the reference implementation i used **fiftyone** to download the dataset and convert it to coco format (probably overkill). And **pycocotools** for mAP evaluation.

evaluation result:
```
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.375
Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.524
Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.405
Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.024
Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.124
Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.414
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.420
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.597
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.623
Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.077
Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.326
Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.675
```

### CODE REVIEW

**Comment by geohot** [comment]
Code looks pretty clean!

Is this meeting the "34.0% mAP" in MLPerf?

**Comment by sohaiberrabii** [comment]
> Code looks pretty clean!
> 
> Is this meeting the "34.0% mAP" in MLPerf? (which of these numbers?)

yes the quality target is 0.34 mAP but the model I used reached 0.3755 mAP: https://github.com/mlcommons/inference/tree/master/vision/classification_and_detection.

**Comment by geohot** [comment]
Testing now. How would you like to be paid? :)

**Comment by geohot** [comment]
Hmm, this `fiftyone` dep is very heavyweight, and broken out of the box on my machine!

```
Subprocess ['/home/tiny/.local/lib/python3.10/site-packages/fiftyone/db/bin/mongod', '--dbpath', '/home/tiny/.fiftyone/var/lib/mongo', '--logpath', '/home/tiny/.fiftyone/var/lib/mongo/log/mongo.log', '--port', '0', '--nounixsocket'] exited with error 127:
/home/tiny/.local/lib/python3.10/site-packages/fiftyone/db/bin/mongod: error while loading shared libraries: libcrypto.so.1.1: cannot open shared object file: No such file or directory
```

It's trying to run mongodb! Can it be removed or replaced by something simpler? `pycocotools` is fine.

**Comment by sohaiberrabii** [comment]
> Hmm, this `fiftyone` dep is very heavyweight (installed 20 other pip packages), and broken out of the box on my machine!
> 
> ```
> Subprocess ['/home/tiny/.local/lib/python3.10/site-packages/fiftyone/db/bin/mongod', '--dbpath', '/home/tiny/.fiftyone/var/lib/mongo', '--logpath', '/home/tiny/.fiftyone/var/lib/mongo/log/mongo.log', '--port', '0', '--nounixsocket'] exited with error 127:
> /home/tiny/.local/lib/python3.10/site-packages/fiftyone/db/bin/mongod: error while loading shared libraries: libcrypto.so.1.1: cannot open shared object file: No such file or directory
> ```
> 
> It's trying to run mongodb! Can it be removed or replaced by something simpler? `pycocotools` is fine.

yeah thats why i said its overkill, im not satisfied with fiftyone at all. 

For download the easiest would be to depend on boto3/botocore or directly call the aws cli (there is a script in mlperf for this https://github.com/mlcommons/inference/blob/master/vision/classification_and_detection/tools/openimages.py).

For conversion to coco, there is also an easy solution https://github.com/bethgelab/openimages2coco.

**Comment by geohot** [comment]
`boto3` dep is fine

**Comment by geohot** [comment]
It's "Downloading images"...

**Comment by geohot** [comment]
Umm...
```
[224/24781] == 106.57 ms loading data, 7.87 ms to run model, 1091.83 ms for postprocessing
[232/24781] == 132.18 ms loading data, 7.93 ms to run model, 1092.62 ms for postprocessing
[240/24781] == 113.84 ms loading data, 7.76 ms to run model, 1120.84 ms for postprocessing
[248/24781] == 131.94 ms loading data, 7.92 ms to run model, 1078.17 ms for postprocessing
[256/24781] == 112.61 ms loading data, 7.74 ms to run model, 1087.59 ms for postprocessing
[264/24781] == 113.55 ms loading data, 7.63 ms to run model, 1098.99 ms for postprocessing
```

What's so slow in postprocessing?

**Comment by sohaiberrabii** [comment]
> ```
> [224/24781] == 106.57 ms loading data, 7.87 ms to run model, 1091.83 ms for postprocessing
> [232/24781] == 132.18 ms loading data, 7.93 ms to run model, 1092.62 ms for postprocessing
> [240/24781] == 113.84 ms loading data, 7.76 ms to run model, 1120.84 ms for postprocessing
> [248/24781] == 131.94 ms loading data, 7.92 ms to run model, 1078.17 ms for postprocessing
> [256/24781] == 112.61 ms loading data, 7.74 ms to run model, 1087.59 ms for postprocessing
> [264/24781] == 113.55 ms loading data, 7.63 ms to run model, 1098.99 ms for postprocessing
> ```
> 
> What's so slow in postprocessing?

In order to filter out low scoring detections:

In [6]: import numpy as np
In [7]: a = np.random.random((8, 120087, 264))

In [8]: %timeit out = a[a>0.05]
1.07 s ± 20.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [9]: a = np.random.random((90000, 264))
In [10]: %timeit out = a[a>0.05]
107 ms ± 4.7 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

Its mostly the first feature map of every image in the batch. Its size is 100 x 100 x num_anchors (=9).

File: examples/mlperf/model_eval.py [hunk 1]
```diff
@@ -42,6 +42,57 @@ def input_fixup(x):
...
+    if dat.shape[0] == bs:
+      outs = mdlrun(dat).numpy()
+    else:
+      mdlrun.jit_cache = None
+      outs =  mdl(input_fixup(dat)).numpy()
+    et = time.perf_counter()
+    predictions = mdl.postprocess_detections(outs, input_size=dat.shape[1:3], orig_image_sizes=[t["image_size"] for t in targets])
+    evaluated_imgs.extend([t["image_id"] for t in targets])
+    coco_results.extend([{"image_id": targets[i]["image_id"], "category_id": label, "bbox": box, "score": score}
+                    for i, prediction in enumerate(predictions) for box, score, label in zip(*prediction.values())])
```
Comment (geohot) [inline examples/mlperf/model_eval.py:84]: Spacing.

Also, isn't this eval O(n^2)? There's no way to do it incrementally?

**Comment by geohot** [comment]
I got this with the last commit

```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.375                                       
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.524                                       
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.405                                       
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.024                                       
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.124                                       
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.414                                       
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.420                                       
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.597                                       
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.623                                       
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.077                                       
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.326                                       
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.675 
 ```
 
 I still don't understand all these numbers. Which one is MLPerf and how do you know?

File: examples/mlperf/model_eval.py (hunk 1 above)
Comment (sohaiberrabii) [reply examples/mlperf/model_eval.py:84]: I might have found a way to do it with pycocotools, i'll test it

**Comment by sohaiberrabii** [comment]
> I got this with the second to last commit
> 
> ```
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.375                                       
>  Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.524                                       
>  Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.405                                       
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.024                                       
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.124                                       
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.414                                       
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.420                                       
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.597                                       
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.623                                       
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.077                                       
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.326                                       
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.675 
> ```
> 
> I still don't understand all these numbers. Which one is MLPerf and how do you know?

They are explained here: cocodataset.org/#detection-evalcocodataset.org/#detection-eval.
The first one corresponds to mlperf quality metric: 
"Metric is COCO box mAP (averaged over IoU of 0.5:0.95)"

**Comment by geohot** [comment]
Ahh cool. So just the spacing and n^2 and I think we're good, I'll test one more time! If there's no way to make the eval non n^2, it's fine to just do it at the end I guess.

**Comment by sohaiberrabii** [comment]
> Ahh cool. So just the spacing and n^2 and I think we're good, I'll test one more time! If there's no way to make the eval non n^2, it's fine to just do it at the end I guess.

after digging through the code of cocoeval, there is indeed no way to compute AP incrementally. At every `accumulate` call, all the detections need to be sorted together. The last commit only reuses the result of `evaluate` (which matches detections to ground truth sin preparation of calling accumulate), this makes the accumulation step at the end faster.

**Comment by geohot** [comment]
Doing the final run now, if it hits I'll merge and the bounty is yours!

How do you want to be paid? Can either post here or e-mail me george@tinygrad.org

**Pull request merged**